File: haskell/src/SimplePrelude.hs
{-# LANGUAGE NoImplicitPrelude #-}

module SimplePrelude
  ( id
  , const
  , (.)
  , P.undefined
  ) where

import qualified Prelude as P

id :: a -> a
id a = a

const :: a -> b -> a
const a _ = a

(.) :: (b -> c) -> (a -> b) -> a -> c
(f . g) x = f (g x)

File: scala/src/main/scala/pg/tc/SimplePrelude.scala
package tc

object SimplePrelude {

  def id[A](a: A): A = a

  def const[A,B](a: A)(b: B): A = a

  def flip[A,B,C](f: A => B => C): B => A => C =
    b => a => f(a)(b)
}
